### PR TITLE
Add REQUEST_OPEN_WELL events to WCONINJH, WCONHIST and valve keywords

### PR DESCRIPTION
**Pull request by vkip**
To ensure shut wells are checked upon changes here.

### CODE REVIEW

**Comment by vkip** [comment]
jenkins build this please

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp [hunk 1]
```diff
@@ -2429,8 +2437,12 @@ Well{0} entered with 'FIELD' parent group:
                     sicd.updateScalingFactor(outlet_segment_length, connections.segment_perf_length(segment_nr));
                 }
 
-                if (well.updateWSEGSICD(sicd_pairs) )
+                if (well.updateWSEGSICD(sicd_pairs) ) {
```
Comment (GitPaean) [inline src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:2440]: It is not desirable to use `well` after `std::move(well)`.  You can avoid that by 
```c++
const bool well_status = well.getStatus();
```

before the line with `std::move()`

It also applies to the two occasions below.

**Comment by GitPaean** [comment]
We assign the event `REQUEST_OPEN_WELL` to `WELOPEN`, `WCONPROD` and `WCONINJE`, because those keywords can update the status for the whole well, potentially open wells are `SHUT` or `STOP` from whatever causes, like SCHEDULE setup, physical causes, economic causes. It was introduced for physical causes and economic causes, which are intended for prediction situations. 

At the same time, I do see historical and prediction WCON* keywords can used in mix in some DATA files, I have no problem to add the event to WCONHIST and WCONINJH. 

But I am not sure about the valve keywords. They are only intended to update the properties for some segments, without touching the status of the whole well. I believe they are different from the WCON* keywords which update the status of the whole well. Do you see some scenarios from the real cases that require Valve keywords can OPEN wells previous SHUT by physical or economic causes?

File: src/opm/input/eclipse/Schedule/KeywordHandlers.cpp (hunk 1 above)
Comment (vkip) [reply src/opm/input/eclipse/Schedule/KeywordHandlers.cpp:2440]: Thanks, will fix.

**Comment by vkip** [comment]
> But I am not sure about the valve keywords. They are only intended to update the properties for some segments, without touching the status of the whole well. I believe they are different from the WCON* keywords which update the status of the whole well. Do you see some scenarios from the real cases that require Valve keywords can OPEN wells previous SHUT by physical or economic causes?

Yes, it could well be that this should be solved in some other manner, but while testing https://github.com/OPM/opm-simulators/pull/4986 I encountered a case where a well is unable to operate with some valves shut, so the well is shut, but once the valves open the well is able to produce.

**Comment by GitPaean** [comment]
> Yes, it could well be that this should be solved in some other manner, but while testing https://github.com/OPM/opm-simulators/pull/4986 I encountered a case where a well (during history) is unable to operate with some valves shut, so the well is shut, but once the valves open the well is able to produce.

For this historical cases,  when you input Valve keywords, I expect you will have a WCONHIST keyword associated with it for that report step to update the rates, right? Is including the event for WCONHIST enough for fixing the problem?

**Comment by steink** [comment]
Sorry for coming in late, probably not relevant for this, but just wanted to make sure https://github.com/OPM/opm-simulators/pull/4986 was run with the most recent commit ([9317929](https://github.com/OPM/opm-simulators/pull/4986/commits/9317929962a7e2c1ef69d7893c54d04e1e9e0257)) which came in a few days after the PR.

**Comment by vkip** [comment]
> Sorry for coming in late, probably not relevant for this, but just wanted to make sure [OPM/opm-simulators#4986](https://github.com/OPM/opm-simulators/pull/4986) was run with the most recent commit ([9317929](https://github.com/OPM/opm-simulators/pull/4986/commits/9317929962a7e2c1ef69d7893c54d04e1e9e0257)) which came in a few days after the PR.

Hi, thanks for checking, but yes, this was run with the latest version.

**Comment by GitPaean** [comment]
Hei, sorry for not getting back sooner, it took some time to think through. 

`REQUEST_OPEN_WELL` was introduced mostly to handle prediction situations. Under prediction situations, wells can be SHUT due to various causes, physical causes (BHP/THP limits, convergence issues, operability issues ) or economic causes. WTEST can be used to test the wells whether they can open again. At the same time, new WELOPEN, WCONPROD or WCONINJE with `OPEN` status can open the wells before WTEST time.  But those are only intended for prediction scenarios, I believe.  The feature of these keywords is that they can OPEN a well previously specified with SHUT status in the DATA file. 

Valve keywords and some other SCHEDULE keywords (WELSEGS, COMPDAT and so on) can update part or whole of the well without touching the STATUS of the well.  If adding them to the event `REQUEST_OPEN_WELL` is doable, while we need to think whether we want this behavoir in the prediction case or WTEST cases. 


Alternatively, since the targeting scenario is a history matching scenario. For history matching scenario, I believe the wells get SHUT due to convergence issue is already not desirable, and well testing facility should NOT be used either. My suggested alternative solution will be, for historical matching wells, we always try to check whether we can reopen the previously SHUT wells in the new time step/report steps, and not depending whether there is an event. 

Let me know what you think.

**Comment by vkip** [comment]
Thanks for thinking about this. After some more pondering at my end also I agree with you - better to always check historic wells. I will update the PR.

At some point I would like to have a broader discussion on possible opening after well change events in general, since I personally find it unnatural not to try to open - but a random PR is not the right place.

**Comment by vkip** [comment]
https://github.com/OPM/opm-simulators/pull/4999

**Pull request closed without merging**